src/cpu/flexcpu/flexcpu.cc

```diff
@@ -0,0 +1,1106 @@
+}
+
+void
+FlexCPU::init()
+{
+    BaseCPU::init();
+
+    if (!params()->switched_out &&
+        system->getMemoryMode() != Enums::timing)
+    {
+        fatal("The Dataflow CPU requires the memory system to be in "
+              "'timing' mode.\n");
+    }
+
+    for (auto& thread : threads) {
+        ThreadContext* tc = thread->getThreadContext();
+        tc->initMemProxies(tc);
+
+        // TODO if fullsystem, we need to check if we need to call initCPU on
+        // the ISA
+    }
+    // TODO add post-construction initialization code
+}
+
+void
+FlexCPU::markActiveCycle()
+{
+    if (curTick() != lastActiveTick) {
+        numCycles++;
+        lastActiveTick = curTick();
+        for (auto& thread : threads) {
+            thread->recordCycleStats();
+        }
+    }
+}
+
+void
+FlexCPU::requestDataAddrTranslation(const RequestPtr& req,
+    ThreadContext* tc, bool write, std::weak_ptr<ExecContext> context,
+    TranslationCallback callback_func)
+{
+    DPRINTF(FlexCPUCoreEvent, "requestDataAddrTranslation()\n");
+    // delay if we're out of slots for simultaneous dtb translations
+
+    Tick queue_time = curTick();
+
+    dataAddrTranslationUnit.addRequest([this, req, tc, write, context,
+                                        callback_func, queue_time]
+    {
+        if (context.expired()) return false; // Don't request for squashed inst
+
+        Tick now = curTick();
+
+        waitingForDataXlation.sample(now - queue_time);
+        // The event callback for when the translation completed comes through
+        // a BaseTLB::Translation object, so we pass that callback straight to
+        // the requestor's callback function by subclassing it.
+        BaseTLB::Translation* handler = new CallbackTransHandler(
+                                            this, callback_func, now, true);
+
+        tc->getDTBPtr()->translateTiming(req, tc, handler,
+                                    write ? BaseTLB::Write : BaseTLB::Read);
+
+        return true;
+    });
+
+    dataAddrTranslationUnit.schedule();
+}
+
+void
+FlexCPU::requestExecution(StaticInstPtr inst,
+                                    weak_ptr<ExecContext> context,
+                                    Trace::InstRecord* trace_data,
+                                    ExecCallback callback_func)
+{
+    DPRINTF(FlexCPUCoreEvent, "requestExecution()\n");
+
+    Tick queue_time = curTick();
+
+    // Do this when we *actually* execute the instruction
+    executionUnit.addRequest([this, inst, context, trace_data, callback_func,
+                              queue_time]
+    {
+        if (context.expired()) return false; // Don't execute squashed inst
+
+        waitingForExecution.sample(curTick() - queue_time);
+
+        Event *e = new EventFunctionWrapper([context, inst, trace_data,
+                                             callback_func]{
+            const shared_ptr<ExecContext> ctxt = context.lock();
+            Fault fault = ctxt ?
```

Comment:
Make this an if statement (I don't like a nested ternary statement). Also, add a comment explaining that the instruction could be squashed while waiting for execution to finish.

Reply:
I'll add the comment definitely.

The nested ternary statement saves an empty object construction + copy/move in this case though... and can be done with predicated execution more easily :T

Reply:
Actually the callback doesn't need to be called if it's expired, so I can just return early if the weak_ptr couldn't be locked.